File: backend/src/main/java/com/autocare/budget/controller/BudgetController.java
package com.autocare.budget.controller;

import com.autocare.budget.dto.BudgetRequestDTO;
import com.autocare.budget.dto.BudgetResponseDTO;
import com.autocare.budget.entity.BudgetStatus;
import com.autocare.budget.service.BudgetService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/budgets")
@RequiredArgsConstructor
public class BudgetController {

    private final BudgetService budgetService;

    @PostMapping
    public ResponseEntity<BudgetResponseDTO> create(@Valid @RequestBody BudgetRequestDTO request) {
        return ResponseEntity.status(HttpStatus.CREATED).body(budgetService.create(request));
    }

    @GetMapping
    public ResponseEntity<List<BudgetResponseDTO>> findAll() {
        return ResponseEntity.ok(budgetService.findAll());
    }

    @GetMapping("/{id}")
    public ResponseEntity<BudgetResponseDTO> findById(@PathVariable String id) {
        return ResponseEntity.ok(budgetService.findByIdResponse(id));
    }

    @GetMapping("/client/{clientId}")
    public ResponseEntity<List<BudgetResponseDTO>> findByClient(@PathVariable String clientId) {
        return ResponseEntity.ok(budgetService.findByClient(clientId));
    }

    @GetMapping("/status/{status}")
    public ResponseEntity<List<BudgetResponseDTO>> findByStatus(@PathVariable BudgetStatus status) {
        return ResponseEntity.ok(budgetService.findByStatus(status));
    }

    @PatchMapping("/{id}/approve")
    public ResponseEntity<BudgetResponseDTO> approve(@PathVariable String id) {
        return ResponseEntity.ok(budgetService.approve(id));
    }

    @PatchMapping("/{id}/reject")
    public ResponseEntity<BudgetResponseDTO> reject(@PathVariable String id) {
        return ResponseEntity.ok(budgetService.reject(id));
    }
}

File: backend/src/budget/budget.controller.ts
import { Body, Controller, Get, HttpCode, HttpStatus, Param, Patch, Post, UseGuards } from "@nestjs/common";
import { BudgetStatus } from "@prisma/client";
import { BudgetService } from "./budget.service";
import { BudgetRequestDto } from "./dto/budget.dto";
import { Roles } from "../common/decorators/roles.decorator";
import { RolesGuard } from "../common/guards/roles.guard";
import { Role } from "@prisma/client";

// Criação/consulta liberada para quem atende o cliente (ADMIN, RECEPTIONIST) e para
// MANAGER (relatórios). Aprovar/rejeitar um orçamento é uma decisão financeira e fica
// restrita a ADMIN e MANAGER.
@Controller("budgets")
@UseGuards(RolesGuard)
@Roles(Role.ADMIN, Role.RECEPTIONIST, Role.MANAGER)
export class BudgetController {
  constructor(private readonly budgetService: BudgetService) {}

  @Post()
  @Roles(Role.ADMIN, Role.RECEPTIONIST)
  @HttpCode(HttpStatus.CREATED)
  create(@Body() dto: BudgetRequestDto) {
    return this.budgetService.create(dto);
  }

  @Get()
  findAll() {
    return this.budgetService.findAll();
  }

  @Get(":id")
  findById(@Param("id") id: string) {
    return this.budgetService.findByIdResponse(id);
  }

  @Get("client/:clientId")
  findByClient(@Param("clientId") clientId: string) {
    return this.budgetService.findByClient(clientId);
  }

  @Get("status/:status")
  findByStatus(@Param("status") status: BudgetStatus) {
    return this.budgetService.findByStatus(status);
  }

  @Patch(":id/approve")
  @Roles(Role.ADMIN, Role.MANAGER)
  approve(@Param("id") id: string) {
    return this.budgetService.approve(id);
  }

  @Patch(":id/reject")
  @Roles(Role.ADMIN, Role.MANAGER)
  reject(@Param("id") id: string) {
    return this.budgetService.reject(id);
  }
}
